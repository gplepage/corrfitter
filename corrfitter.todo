- Add a median() function (like in lsqfit.Accumulator) to Dataset.

- Fix assemble(): doesn't work now if template contains reg-expression symbols (like +) since the template elements are used as regular expressions in one place (grep()) but later as literal strings. Must be all the same.