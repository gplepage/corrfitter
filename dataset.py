#!/usr/bin/env python
# encoding: utf-8
"""
dataset.py --- simplified replacement for old module; for legacy purposes only 
               (use gvar.dataset for new stuff).
"""
# Created by G. Peter Lepage, Cornell University, on 2012-05-22.
# Copyright (c) 2010-2012 G. Peter Lepage.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version (see <http://www.gnu.org/licenses/>).
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

import gvar
import numpy

def flexargfcn(f):
    """ decorator for Dataset analysis functions """
    def newf(self,arg=None):
        if arg is not None:
            if (hasattr(arg,"__hash__") and arg.__hash__ is not None 
                and arg in self): 
                return f(self,self[arg])
            else:
                ans = {}
                for k in arg:
                    ans[k] = f(self,self[k]) 
                return ans
        else:
            ans = {}
            for k in self:
                ans[k] = f(self,self[k]) 
            return ans
    ##
    newf.__doc__ = f.__doc__
    return newf
##

class Dataset(gvar.dataset.Dataset):
    """ legacy class --- don't use otherwise """
    def __init__(self, *args,**kargs):
        if 'bstrap' in kargs:
            bstrap = kargs['bstrap']
            del kargs['bstrap']
        else:
            bstrap = False
        super(Dataset, self).__init__(*args,**kargs)
        self.bstrap = bstrap
    ##
    def copy(self,dlist):
        if type(dlist) not in [list,tuple]:
            dlist = [dlist]
        for d in dlist:
            self.extend(d)
    ##
    def nmeas(self):
        dlen = [len(self[k]) for k in self]
        return min(dlen),max(dlen)
    ##
    @flexargfcn
    def avg(self,x):
        return numpy.mean(x,axis=0)
    ##
    @flexargfcn
    def median(self,x):
        return gvar.dataset.avg_data(x,median=True,spread=self.bstrap)
    ##
    @flexargfcn
    def sdev(self,x):
        if self.bstrap:
            return numpy.std(xx,axis=0)
        else:
            return numpy.std(xx,axis=0)/math.sqrt(len(xx))
    ##
    @flexargfcn
    def cov(self,d):
        oldshape = d[0].shape
        d = [di.flatten() for di in d]
        # d = d.reshape((d.shape[0],-1)) # flatten all other dimensions
        dcov = numpy.cov(d,rowvar=False,bias=True)
        if self.bstrap:
            return dcov.reshape(oldshape+oldshape)
        else:
            return dcov.reshape(oldshape+oldshape)/float(len(d))
    ##     
    def gdev(self,arg=None):
        if arg is None:
            arg = self.keys()
            strip_ans = False
        elif (hasattr(arg,'__hash__') and arg.__hash__ is not None 
              and arg in self):
            arg = [arg]
            strip_ans = True
        if len(arg)==0:
            return dict()
        dd = dict()
        for k in arg:
            dd[k] = self[k]
        ans = gvar.dataset.avg_data(dd,spread=self.bstrap)
        return ans[arg[0]] if strip_ans else ans
    ##
    def bin(self,nbin=2):
        return gvar.bin_data(self,binsize=nbin)
    ##
    def bootstrap_iter(self,n=None):
        return gvar.dataset.bootstrap_iter(self,n)
    ##
    def assemble(self,template,newtag=''):
        ans = Dataset()
        ans[newtag] = self.arrayzip(template)
        return ans
##  

def tabulate_avg(avgout,format=(6,3)):
    """ Tabulates averages and standard deviations.
        
    tabulate_avg(...) creates a nicely formatted table displaying the
    output from functions like ``dataset.Dataset.gdev``. Here ``avgout`` is
    the output. Parameter ``format`` specifies the output format:
    ``format=(N,D)`` implies that format ``'%N.Df(%Dd)'`` is used to print
    ``avg,int(10**D * std_dev)``. The table is returned as a single string,
    for printing.
    """
    table = []
    output = avgout.items()
    output.sort()
    for tag,avsd in output:
        try:
            av = avsd.mean
            sd = avsd.sdev
        except AttributeError:
            av = gvar.mean(avsd)
            sd = gvar.sdev(avsd)
        lines = ''
        line = '%15s' % str(tag)
        try:
            sdfac = 10**format[1]
            fmt = (' %'+str(format[0])+'.'+str(format[1])+
                  'f(%'+str(format[1])+'d)')
            def avgfmt(av,sd,fmt=fmt,sdfac=sdfac):
                try:
                    return fmt % (av,int(sdfac*sd+0.5))
                except:
                    return (' %g (%.4g)' % (av,sd))
            ##
        except:
            def avgfmt(av,sd):
                return (' %g (%.4g)' % (av,sd))
            ##
        na = len(av)
        if len(sd)<na:
            na = len(sd)
        if na>=1:
            for i in xrange(na):
                if len(sd.shape)==2:
                    sdi = math.sqrt(sd[i][i])
                else:
                    sdi = sd[i]
                nextfield = avgfmt(av[i],sdi)
                if (len(nextfield)+len(line))>78:
                    lines = lines + line + '\n'
                    line = ''.ljust(15) + nextfield
                else:
                    line = line + nextfield
            table.append(lines + line +'\n')
    return '\n'.join(table)
##    
    

if __name__ == '__main__':
    import gvar
    gvar.ranseed((1950,1))
    r1 = gvar.gvar(8.,1.)
    r2 = gvar.gvar([-10.,-9.],[2.,3.])
    r3 = gvar.gvar([[0.,1.],[2.,3.]],[[1.,2.],[3.,4.]])
    r3_iter = gvar.raniter(r3)
    r2_iter = gvar.raniter(r2)
    N = 1001
    d = Dataset(bstrap=False)
    for x in range(N):
        d.append('x',r3_iter.next())
    for x in range(N):
        d.append('y',r2_iter.next())
    d2 = Dataset(bstrap=False)
    for x in range(N):
        d2.append('z',r1())
    d.copy(d2)
    med = d.gdev()
    for k in med:
        print( k,med[k])
    avg = d.avg()
    for k in avg:
        print( k,avg)
    print( d.nmeas())
    nd = d.assemble(['y','y'],'yy')
    print( nd.avg())
    nd = d.grep('x|y')
    print( nd.keys())
        
    output = """    
    y [-10.0069 +- 0.00632473 -8.98159 +- 0.00949937]
    x [[-0.00311314 +- 0.00316463 1.00027 +- 0.00634238]
     [2.00083 +- 0.00950991 2.99893 +- 0.0126476]]
    z 8.00641 +- 0.00315755
    """     # 15.5 sec with N = 100001