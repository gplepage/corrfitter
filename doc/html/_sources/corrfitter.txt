:mod:`corrfitter` - Least-Squares Fit to Correlators
=====================================================

.. moduleauthor:: G.P. Lepage <g.p.lepage@cornell.edu>

.. |CorrFitter| replace:: :class:`corrfitter.CorrFitter`
.. |Corr2| replace:: :class:`corrfitter.Corr2`
.. |Corr3| replace:: :class:`corrfitter.Corr3`
.. |BaseModel| replace:: :class:`corrfitter.BaseModel`
.. |Dataset| replace:: :class:`gvar.dataset.Dataset`
.. |GVar| replace:: :class:`gvar.GVar`

.. automodule:: corrfitter
   :synopsis: Least-squares fitter for correlators.


Correlator Model Objects
-------------------------
Correlator objects describe theoretical models that are fit to
correlator data by varying the models' parameters. 

A model object's parameters are specified through priors for the fit. A
model assigns labels to each of its parameters (or arrays of related
parameters), and these labels are used to identify the corresponding
parameters in the prior. Parameters can be shared by more than one model
object.

A model object also specifies the data that it is to model. The data is 
identified by the data tag that labels it in the input file or |Dataset|.

.. autoclass:: corrfitter.Corr2
   :members:
   
.. autoclass:: corrfitter.Corr3
   :members:
   
.. autoclass:: corrfitter.BaseModel
   :members:


|CorrFitter| Objects
---------------------
|CorrFitter| objects are wrappers for :func:`lsqfit.nonlinear_fit()` which
is used to fit a collection of models to a collection of Monte Carlo data.

.. autoclass:: corrfitter.CorrFitter
   :members:

