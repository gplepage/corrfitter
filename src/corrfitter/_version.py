__version__ = '8.2'