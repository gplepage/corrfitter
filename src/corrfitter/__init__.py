"""
This module contains tools that facilitate least-squares fits, as functions
of time ``t``, of simulation (or other statistical) data for 2-point and
3-point correlators of the form::

    Gab(t)    =  <b(t) a(0)>
    Gavb(t,T) =  <b(T) V(t) a(0)>

where ``T > t > 0``. Each correlator is modeled using |Corr2| for 2-point
correlators, or |Corr3| for 3-point correlators in terms of amplitudes for
each source ``a``, sink ``b``, and vertex ``V``, and the energies
associated with each intermediate state. The amplitudes and energies are
adjusted in the least-squares fit to reproduce the data; they are defined
in a shared prior (typically a dictionary).

An object of type |CorrFitter| describes a collection of correlators and is
used to fit multiple models to data simultaneously. Fitting multiple
correlators simultaneously is important if there are statistical
correlations between the correlators. Any number of correlators may be
described and fit by a single |CorrFitter| object.

Typical code has the structure ::

    from corrfitter import CorrFitter

    data = make_data('mcfile')          # user-supplied routine
    models = make_models()              # user-supplied routine
    N = 4                               # number of terms in fit functions
    prior = make_prior(N)               # user-supplied routine
    fitter = CorrFitter(models=models)
    fit = fitter.lsqfit(data=data, prior=prior)  # do the fit
    print_results(fit, prior, data)     # user-supplied routine

where ``make_data`` assembles the correlator data to be fit,
``make_prior`` defines Bayesian priors for the fit parameters,
``fitter.lsqfit(...)`` does the fit, and ``print_results``
writes out the rsults. Sample code for each of these routines is
given in the tutorial documentation for |CorrFitter|.
"""

# Created by G. Peter Lepage, Cornell University, on 2010-11-26.
# Copyright (c) 2010-2021 G. Peter Lepage.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version (see <http://www.gnu.org/licenses/>).
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

from ._corrfitter import * 
from ._version import __version__