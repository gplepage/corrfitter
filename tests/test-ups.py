from corrfitter import Corr2,CorrFitter
from lsqfit import wavg
from gvar import gvar,log,exp,evalcov,mean,sdev,BufferDict
from gvar.dataset import Dataset,avg_data
import lsqfit
import gvar as gd
import time
import numpy
from numpy import linalg

lsqfit.nonlinear_fit.fmt_parameter = '%7.3f +- %7.3f'

NTERM_PRIOR = 8             # number of terms in prior
MC = None
RATIO = True

USE_MARGINALIZATION = False  # run mode
    
DISPLAY_PLOTS = False       # display plots at end
try: 
    import matplotlib
except ImportError:
    DISPLAYPLOTS = False

def main():
    dfile = 'ups.bin24'     # input data file
    svdcut = 1e-3          # needed even without marginalization
    svdnum = 113            # number of samples in ups.bin24
    data = avg_data(Dataset(dfile))
    # for k in data:
    #     print k,gd.is_independent(data[k])[0]
    # c = gd.evalcov(data)
    models = make_models()
    if USE_MARGINALIZATION:
        prior = make_prior(NTERM_PRIOR)
    starttime = time.clock()
    p0 = None
    ntermlist = [4] if USE_MARGINALIZATION else [8]
    for nterm in ntermlist:
        if not USE_MARGINALIZATION:
            prior = make_prior(nterm)
        fitter = CorrFitter(models=models,svdcut=svdcut,
                            nterm=nterm if USE_MARGINALIZATION else None,
                            mc=MC,ratio=RATIO,maxit=10000)
        print 30*'=','nterm =',nterm,'  nterm_prior =',len(prior['logdE']),
        print '   MC =',MC,RATIO
        fit = fitter.lsqfit(prior=prior,data=data,p0=p0,svdnum=svdnum)
        # print fit
        print_results(fitter,prior,data)
        if nterm==1:
            print_effmass(fitter,prior,data)
        if not isinstance(p0,str):      # not using a file?
            print '\nusing last fit\n'
            p0 = fit.pmean
        print '\nt=%.1f ====' % (time.clock()-starttime)
        print
    if DISPLAY_PLOTS:
        fitter.display_plots()
##

def print_effmass(fitter,prior,data):
    meff = []
    xdata = fitter.fit.data[1]
    print '\neff mass:'
    for m in fitter.models:
        G = xdata[m.datatag]
        meff.append(wavg(log(G[:-1]/G[1:])))
        print '%10s:'%m.datatag,meff[-1],'  chi2/dof = %.2f  Q = %.2f'%(wavg.chi2/wavg.dof,wavg.Q)
    print '\n%10s:'%'all not d',wavg(meff[:-4]),'  chi2/dof = %.2f  Q = %.2f'%(wavg.chi2/wavg.dof,wavg.Q)
##

def print_results(fitter,prior,data):
    fit = fitter.fit
    p = fit.p
    dE = exp(p['logdE'])
    E = [sum(dE[:i+1])-(dE[0] if i!=0 else 0.0) for i in range(len(dE))]
    print '   E =',fmtlist(E),'  (En-E0 for n>0)'
    if 'logl' in p:
        print '   l =',fmtlist(exp(p['logl']))
    for k in ['g','d','e']:
        if k in p:
            print '   %s ='%k,fmtlist(p[k])
    print
    ## error budget ##
    inputs = dict(stat=[data[k] for k in data])
    inputs.update(prior)
    outputs = dict(E0=E[0])
    if len(E)>1:
        outputs['dE1'] = E[1]
    print fit.fmt_partialsdev(outputs,inputs,ndigit=3)
    ##
    print '\ncorrelators:\n   ',[m.datatag for m in fitter.models]
##
        
def make_prior(nterm):
    prior = BufferDict()
    c0 = 0.1
    dc = 1.0
    prior.add('logl',[log(gvar(0.4,0.4)) for i in range(nterm)])
    prior.add('g',numpy.array([gvar(c0,dc) for i in range(nterm)]))
    prior.add('e',[gvar(c0,dc) for i in range(nterm)])
    prior.add('d',[gvar(c0,dc) for i in range(nterm)])
    prior.add('logdE',log([gvar(0.5,0.5) for i in range(nterm)]))
    prior['logdE'][0] = log(gvar(.256,0.1))
    return prior
##

def make_models():
    # tfit = [1,2,3,5,8,13,21]    # fibonacci numbers
    tfit = range(1,24)          # full range
    models = [
    Corr2('1s0.ll',a='logl',b='logl',dE='logdE',tdata=range(1,24),tfit=tfit)
    ,
    Corr2('1s0.lg',a='logl',b='g',dE='logdE',
        tdata=range(1,24),tfit=tfit,othertags=['1s0.gl'])
    ,
    Corr2('1s0.gg',a='g',b='g',dE='logdE',tdata=range(1,24),tfit=tfit)
    ,
    Corr2('1s0.le',a='logl',b='e',dE='logdE',
        tdata=range(1,24),tfit=tfit,othertags=['1s0.el'])
    ,
    Corr2('1s0.ge',a='g',b='e',dE='logdE',
        tdata=range(1,24),tfit=tfit,othertags=['1s0.eg'])
    ,
    Corr2('1s0.ee',a='e',b='e',dE='logdE',tdata=range(1,24),tfit=tfit)
    ,
    Corr2('1s0.ld',a='logl',b='d',dE='logdE',
        tdata=range(1,24),tfit=tfit,othertags=['1s0.dl'])
    ,
    Corr2('1s0.gd',a='g',b='d',dE='logdE',
        tdata=range(1,24),tfit=tfit,othertags=['1s0.dg'])
    ,
    Corr2('1s0.ed',a='e',b='d',dE='logdE',
        tdata=range(1,24),tfit=tfit,othertags=['1s0.de'])
    ,
    Corr2('1s0.dd',a='d',b='d',dE='logdE',tdata=range(1,24),tfit=tfit)
    ]
    return models[:]
##

def fmtlist(x):
    """ Make neat lists for printing. """
    return '  '.join([xi.fmt() for xi in x])
##
    
if __name__ == '__main__':
    if True:
        main()
    else:
        import hotshot, hotshot.stats
        prof = hotshot.Profile("lsqfit-test1.prof")
        prof.runcall(main)
        prof.close()
        stats = hotshot.stats.load("lsqfit-test1.prof")
        stats.strip_dirs()
        stats.sort_stats('time', 'calls')
        stats.print_stats(40)
