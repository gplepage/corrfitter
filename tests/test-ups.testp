(dp0
S'logdE'
p1
cnumpy.core.multiarray
_reconstruct
p2
(cnumpy
ndarray
p3
(I0
tp4
S'b'
p5
tp6
Rp7
(I1
(I8
tp8
cnumpy
dtype
p9
(S'f8'
p10
I0
I1
tp11
Rp12
(I3
S'<'
p13
NNNI-1
I-1
I0
tp14
bI00
S'\xe1\x0bl\xd8\x00\xc9\xf5\xbf\n%7\xa6\x8b\x08\xe5\xbfZ\x1064b\x08\xf6\xbf\xf0\xbc\xa1\x13~\x04\xe9\xbfx\xff\xda\xf3j\xf5\xf0\xbf\xbd\xb8\xd4\xd8\xccx\xe7\xbfEu\x1c\xd1b\xa9\xe0\xbf\xee\xdd\xc1\x81B\xc1\xe0\xbf'
p15
tp16
bsS'logl'
p17
g2
(g3
(I0
tp18
g5
tp19
Rp20
(I1
(I8
tp21
g12
I00
S'\x8a\x9b`\\5\xcb\xe5\xbf]q\xbf\xacF\xdf\xec\xbf\xb9\xf3\xc2\xc7\x85\x8a\xf5\xbf\x92#\xbb\xf8\x18A\xee\xbf^\xeb\xc4/\xd2o\xf2\xbf\xbe\x12\xb9)~\x02\xf3\xbf\x8b\x9d}Vr\xbb\xf4\xbf\x84l>\xf3y\xb2\xf4\xbf'
p22
tp23
bsS'e'
p24
g2
(g3
(I0
tp25
g5
tp26
Rp27
(I1
(I8
tp28
g12
I00
S'\xdf\xce\xde\x9c\x92;\xc9?r\x9b\xd7U#\xa1\xd4?#z\x1d\n\x93\x9f\xd1\xbf\xbc\x96]\xbe\x83M\xc9?\x83\xf5\x94I\xedU\xd4\xbfCb\x8e\x19\x8f\xfc<?\x81\xde\xa0A#\xfeo?\xa4\xd1j\xddVR\xd4?'
p29
tp30
bsS'd'
p31
g2
(g3
(I0
tp32
g5
tp33
Rp34
(I1
(I8
tp35
g12
I00
S'\xa2\xbc\xf2a$\xb1\xcb?\x98j+\xf8D\x9a\xd8?\xf9KYuT\xe7\xd2\xbf\xa8\xb1\xab{\xf8\xc6\x8a\xbf\tJ\xac\xc3\xbe\xe7\xab\xbf\xf2i\x9d\x9b\xdc\x01\xb9?\xd9\xcee^\xce\xbf\xb4?3\xa7i~\xe8\\\xe3?'
p36
tp37
bsS'g'
p38
g2
(g3
(I0
tp39
g5
tp40
Rp41
(I1
(I8
tp42
g12
I00
S'T\xeeN\xad\x08\xd6\xeb?\xd6\xa0X\xa6\xd4\x94\xc2\xbf7\x95\xbew\xc9\x05\x98?\xd7<\xee8\xfc\x19\xc2?e\x95\x91\xb1o,\xc5?1\xadh\xc0\x16\x11\xc3\xbf\x7fh\x80\x0c\xb3\x9b\x9b\xbf\x19\xa4\x00?e\x07\xc1?'
p43
tp44
bs.