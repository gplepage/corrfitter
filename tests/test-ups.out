============================== nterm = 8   nterm_prior = 8    RATIO = True
Least Square Fit:
  chi2/dof [dof] = 0.9 [113]    Q = 0.76    logGBF = -192.47    itns = 1

Parameters:
         logl 0   -0.6811 (18)      [ -0.9 (1.0) ]
              1    -0.902 (32)      [ -0.9 (1.0) ]
              2     -1.35 (49)      [ -0.9 (1.0) ]
              3     -0.95 (38)      [ -0.9 (1.0) ]
              4     -1.15 (71)      [ -0.9 (1.0) ]
              5     -1.19 (79)      [ -0.9 (1.0) ]
              6     -1.30 (91)      [ -0.9 (1.0) ]
              7     -1.29 (88)      [ -0.9 (1.0) ]
            g 0    0.8699 (17)      [  0.1 (1.0) ]
              1    -0.145 (11)      [  0.1 (1.0) ]
              2     0.023 (33)      [  0.1 (1.0) ]
              3      0.14 (10)      [  0.1 (1.0) ]
              4      0.17 (19)      [  0.1 (1.0) ]
              5     -0.15 (33)      [  0.1 (1.0) ]
              6     -0.03 (87)      [  0.1 (1.0) ]
              7      0.13 (39)      [  0.1 (1.0) ]
            e 0   0.19713 (46)      [  0.1 (1.0) ]
              1     0.322 (26)      [  0.1 (1.0) ]
              2    -0.275 (69)      [  0.1 (1.0) ]
              3      0.20 (17)      [  0.1 (1.0) ]
              4     -0.32 (17)      [  0.1 (1.0) ]
              5      0.00 (60)      [  0.1 (1.0) ]
              6      0.00 (83)      [  0.1 (1.0) ]
              7      0.32 (38)      [  0.1 (1.0) ]
            d 0   0.21634 (47)      [  0.1 (1.0) ]
              1     0.384 (26)      [  0.1 (1.0) ]
              2    -0.295 (42)      [  0.1 (1.0) ]
              3    -0.013 (93)      [  0.1 (1.0) ]
              4     -0.05 (18)      [  0.1 (1.0) ]
              5      0.10 (42)      [  0.1 (1.0) ]
              6      0.08 (85)      [  0.1 (1.0) ]
              7      0.61 (29)      [  0.1 (1.0) ]
        logdE 0   -1.3616 (10)      [ -1.36 (39) ]
              1    -0.657 (27)      [ -0.7 (1.0) ]
              2     -1.38 (29)      [ -0.7 (1.0) ]
              3     -0.78 (37)      [ -0.7 (1.0) ]
              4     -1.06 (88)      [ -0.7 (1.0) ]
              5     -0.73 (96)      [ -0.7 (1.0) ]
              6     -0.52 (97)      [ -0.7 (1.0) ]
              7     -0.52 (98)      [ -0.7 (1.0) ]

Settings:
  svdcut = (0.001,None)   svdnum = (1.1e+02,None)    reltol/abstol = 1e-10/1e-10

   E = 0.25626(26)  0.518(14)  0.771(80)  1.23(19)  1.57(30)  2.05(54)  2.65(77)  3.24(93)   (En-E0 for n>0)
   l = 0.50608(94)  0.406(13)  0.26(13)  0.39(15)  0.32(22)  0.30(24)  0.27(25)  0.27(24)
   g = 0.8699(17)  -0.145(11)  0.023(33)  0.14(10)  0.17(19)  -0.15(33)  -0.03(87)  0.13(39)
   d = 0.21634(47)  0.384(26)  -0.295(42)  -0.013(93)  -0.05(18)  0.10(42)  0.08(85)  0.61(29)
   e = 0.19713(46)  0.322(26)  -0.275(69)  0.20(17)  -0.32(17)  0.00(60)  0.00(83)  0.32(38)

Partial % Errors:
                           dE1        E0
----------------------------------------
              logdE:     0.474     0.011
                  e:     0.490     0.013
                  d:     0.229     0.004
                  g:     0.377     0.007
               stat:     1.895     0.091
               logl:     1.116     0.015
----------------------------------------
              total:     2.664     0.101


correlators:
    ['1s0.ll', '1s0.lg', '1s0.gg', '1s0.le', '1s0.ge', '1s0.ee', '1s0.ld', '1s0.gd', '1s0.ed', '1s0.dd']

using last fit


t=0.8 ====

